runtime/near-vm-runner/src/wasmer2_runner.rs

```diff
@@ -266,26 +272,106 @@ impl Wasmer2VM {
-            .map_err(|e| CompilationError::WasmerCompileError { msg: e.to_string() })
+        let executable = self
+            .engine
+            .compile_universal(&prepared_code, &self)
+            .map_err(|e| CompilationError::WasmerCompileError { msg: e.to_string() })?;
+        Ok(executable)
     }
 
-    pub(crate) unsafe fn deserialize(
+    pub(crate) fn compile_and_load(
         &self,
-        serialized: &[u8],
-    ) -> Result<VMArtifact, DeserializeError> {
-        let executable = UniversalExecutableRef::deserialize(serialized)?;
-        let artifact = self
-            .engine
-            .load_universal_executable_ref(&executable)
-            .map_err(|e| DeserializeError::Compiler(e))?;
-        Ok(Arc::new(artifact))
+        code: &ContractCode,
+        cache: Option<&dyn CompiledContractCache>,
+    ) -> Result<Result<VMArtifact, CompilationError>, CacheError> {
+        let key = get_contract_cache_key(code, VMKind::Wasmer2, &self.config);
+
+        let compile_or_read_from_cache =
+            || -> Result<Result<VMArtifact, CompilationError>, CacheError> {
+                let cache_record = cache
+                    .map(|cache| cache.get(&key))
+                    .transpose()
+                    .map_err(|_: io::Error| CacheError::ReadError)?
+                    .flatten();
+
+                let stored_artifact: Option<VMArtifact> = match cache_record {
+                    None => None,
+                    Some(CompiledContract::CompileModuleError(err)) => return Ok(Err(err)),
+                    Some(CompiledContract::Code(serialized_module)) => {
+                        unsafe {
+                            // (UN-)SAFETY: the `serialized_module` must have been produced by a prior call to
+                            // `serialize`.
+                            //
+                            // In practice this is not necessarily true. One could have forgotten to change the
+                            // cache key when upgrading the version of the wasmer library or the database could
+                            // have had its data corrupted while at rest.
+                            //
+                            // There should definitely be some validation in wasmer to ensure we load what we think
+                            // we load.
+                            let executable =
+                                UniversalExecutableRef::deserialize(&serialized_module)
+                                    .map_err(|_| CacheError::DeserializationError)?;
+                            let artifact = self
+                                .engine
+                                .load_universal_executable_ref(&executable)
+                                .map(Arc::new)
+                                .map_err(|_| CacheError::DeserializationError)?;
+                            Some(artifact)
+                        }
+                    }
+                };
```

Comment:
This pattern seems to be shared between VMs and feels somewhat error-prone. Can this be abstracted into a function?

Reply:
Yeah, it seems fine to not roll up that into this PR at least...

Reply:
Ah, I see, yeah, we can I guess add some helper to CompiledContractCache, will note in the relevant issue